Source/WebCore/rendering/RenderTheme.cpp

```diff
@@ -456,6 +457,189 @@ ControlPartType RenderTheme::autoAppearanceForElement(RenderStyle& style, const
     return ControlPartType::NoControl;
 }
 
+static RefPtr<ControlPart> creatMeterPartForRenderer(const RenderObject& renderer)
+{
+    ASSERT(is<RenderMeter>(renderer));
+    const auto& renderMeter = downcast<RenderMeter>(renderer);
+
+    auto element = renderMeter.meterElement();
+    MeterPart::GaugeRegion gaugeRegion;
+
+    switch (element->gaugeRegion()) {
+    case HTMLMeterElement::GaugeRegionOptimum:
+        gaugeRegion = MeterPart::GaugeRegion::Optimum;
+        break;
+    case HTMLMeterElement::GaugeRegionSuboptimal:
+        gaugeRegion = MeterPart::GaugeRegion::Suboptimal;
+        break;
+    case HTMLMeterElement::GaugeRegionEvenLessGood:
+        gaugeRegion = MeterPart::GaugeRegion::EvenLessGood;
+        break;
+    }
+
+    return MeterPart::create(gaugeRegion, element->value(), element->min(), element->max());
+}
+
+RefPtr<ControlPart> RenderTheme::createControlPartForRenderer(const RenderObject& renderer) const
+{
+    ControlPartType type = renderer.style().effectiveAppearance();
+
+    switch (type) {
+    case ControlPartType::NoControl:
+    case ControlPartType::Auto:
+        break;
+
+    case ControlPartType::Checkbox:
+    case ControlPartType::Radio:
+    case ControlPartType::PushButton:
+    case ControlPartType::SquareButton:
+    case ControlPartType::Button:
+    case ControlPartType::DefaultButton:
+    case ControlPartType::Listbox:
+    case ControlPartType::Menulist:
+    case ControlPartType::MenulistButton:
+        break;
+
+    case ControlPartType::Meter:
+        return creatMeterPartForRenderer(renderer);
+
+    case ControlPartType::ProgressBar:
+    case ControlPartType::SliderHorizontal:
+    case ControlPartType::SliderVertical:
+    case ControlPartType::SearchField:
+#if ENABLE(APPLE_PAY)
+    case ControlPartType::ApplePayButton:
+#endif
+#if ENABLE(ATTACHMENT_ELEMENT)
+    case ControlPartType::Attachment:
+    case ControlPartType::BorderlessAttachment:
+#endif
+    case ControlPartType::TextArea:
+    case ControlPartType::TextField:
+    case ControlPartType::CapsLockIndicator:
+#if ENABLE(INPUT_TYPE_COLOR)
+    case ControlPartType::ColorWell:
+#endif
+#if ENABLE(SERVICE_CONTROLS)
+    case ControlPartType::ImageControlsButton:
+#endif
+    case ControlPartType::InnerSpinButton:
+#if ENABLE(DATALIST_ELEMENT)
+    case ControlPartType::ListButton:
+#endif
+    case ControlPartType::SearchFieldDecoration:
+    case ControlPartType::SearchFieldResultsDecoration:
+    case ControlPartType::SearchFieldResultsButton:
+    case ControlPartType::SearchFieldCancelButton:
+    case ControlPartType::SliderThumbHorizontal:
+    case ControlPartType::SliderThumbVertical:
+        break;
+    }
+
+    ASSERT_NOT_REACHED();
+    return nullptr;
+}
+
+static auto& controlPartRendererMap()
+{
+    using ControlPartRendererMap = HashMap<const RenderObject*, Ref<ControlPart>>;
```

Comment:
Should we store references to ControlParts in `RenderObjectRareData` instead?

Reply:
Done. A new data member of type `RefPtr<ControlPart>` was added to `RenderObjectRareData`.